### PR TITLE
feat: support route in cluster mode

### LINKED ISSUE #92: Start CeresDB in distributed mode with CeresMeta
**Description**
The target is to run ceresdb in distributed mode and in order for that ceresdb in distributed mode, ceresdb must communicate with ceresmeta by the meta client. 

**Proposal**
CeresMeta's service is built by gRPC and the corresponding protocol has been defined by [the proto file](https://github.com/CeresDB/ceresdbproto/blob/main/protos/meta/service.proto) so first what we need to do is build a grpc client based on that proto file during ceresdb's starting procedure.

And there is no need to keep the old meta client, and the new meta client should provide the capabilities:
- Send keepalive heartbeats to ceresmeta;
- Register handlers for node heartbeats;
- Allocate schema/table id;
- Provide routing information of the cluster from the ceresmeta;

And based on the new meta client (meta_client_v2) a module called cluster will be responsible for managing information from the CeresMeta, including:
- Creating schema, creating/deleting tables;
- Tables to be opened in the ceresdb instance;
- Events such as change of the cluster topology from ceresmeta;
- ...

**Additional context**
steps:
- [x] Upgrade [ceresdbproto](https://github.com/CeresDB/ceresdbproto), remove old meta client and refactor meta_client_v2 (#100)
- [x] Enhance cluster module based on meta_client_v2 (#100)
- [x] Refactor volatile catalog implementation (#157)
- [x] Support startup in distributed mode (#190)
- [x] Remove the old meta_client (#205)
- [ ] Provide routing information (#215)

### PR DESCRIPTION
**Pull request by ShiKaiWi**
# Which issue does this PR close?

Closes #92 

# Rationale for this change
 The routing information when starting CeresDB with CeresMeta is stored in CeresMeta, and we need to provide such information to the clients by `route` grpc service.

# What changes are included in this PR?
Implement the router based on `Cluster`, and then provide the `route` gRPC service based on the router.

# Are there any user-facing changes?
None because the `route` service is compatible.

# How does this change test
Some new unit tests to ensure this PR.

### CODE REVIEW

File: server/src/schema_config_provider/cluster_based.rs [hunk 1]
```diff
@@ -4,22 +4,28 @@
 
 #![allow(dead_code)]
```
Comment (waynexia) [inline server/src/schema_config_provider/cluster_based.rs:5]: What will happen if we remove this

File: server/src/route/cluster_based.rs [hunk 2]
```diff
@@ -19,8 +26,58 @@ impl ClusterBasedRouter {
     }
 }
 
+// Parse the raw endpoint which should be the form: <domain_name>:<port>
+//
+// Returns `None` if fail to parse.
+fn try_parse_endpoint(raw: &str) -> Option<Endpoint> {
```
Comment (waynexia) [inline server/src/route/cluster_based.rs:32]: standard library has this trait https://doc.rust-lang.org/std/str/trait.FromStr.html#tymethod.from_str (which provides `.parse()`)

File: server/src/config.rs [hunk 3]
```diff
@@ -46,6 +46,12 @@ pub struct StaticRouteConfig {
     pub topology: StaticTopologyConfig,
 }
 
+#[derive(Debug, Clone, Deserialize)]
+pub struct Node {
+    pub addr: String,
+    pub port: u16,
+}
```
Comment (waynexia) [inline server/src/config.rs:53]: This is the same with `EndPoint`, do we have any way to reduce different naming on one thing?

File: server/src/schema_config_provider/cluster_based.rs [hunk 4]
```diff
@@ -4,22 +4,28 @@
...
 
-use cluster::ClusterRef;
+use cluster::{topology::SchemaConfig, ClusterRef};
 
 use crate::schema_config_provider::{Result, SchemaConfigProvider};
 
 pub struct ClusterBasedProvider {
     cluster: ClusterRef,
+    default_schema_config: SchemaConfig,
```
Comment (waynexia) [inline server/src/schema_config_provider/cluster_based.rs:13]: According to the `FIXME` beneath this `default_schema_config` is also temporary, right?

File: server/src/route/cluster_based.rs [hunk 5]
```diff
@@ -19,8 +26,58 @@ impl ClusterBasedRouter {
...
+    let (domain, raw_port) = raw.split_once(':')?;
+    let port: u16 = raw_port.parse().map(Some).unwrap_or(None)?;
```
Comment (waynexia) [inline server/src/route/cluster_based.rs:34]:
```suggestion
    let port: u16 = raw_port.parse().ok()?;
```

File: server/src/route/cluster_based.rs [hunk 6]
```diff
@@ -19,8 +26,58 @@ impl ClusterBasedRouter {
...
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(ErrWithCause {
+                code: StatusCode::InternalError,
+                msg: "Fail to route tables by cluster",
+            })?;
+
+        let mut routes = Vec::with_capacity(route_resp.entries.len());
+
+        // Now we pick up the nodes who own the leader shard for the route response.
```
Comment (waynexia) [inline server/src/route/cluster_based.rs:61]: Is leader only here a hardcode? Maybe we need to add one field in `RouteRequest` to let the client specify what role it wants. Like `LeaderOnly`, `InSync`, `NoSync`, `DontCare` etc.

File: server/src/config.rs (hunk 3 above)
Comment (ShiKaiWi) [reply server/src/config.rs:53]: `Endpoint` is actually defined in `ceresdbproto` so we can't use it as some item of the configuration.

File: server/src/route/cluster_based.rs (hunk 6 above)
Comment (ShiKaiWi) [reply server/src/route/cluster_based.rs:61]: It's true, but for now the protocol hasn't supported such things yet. Let's add the information about roles afterwards when the protocol upgraded.

File: server/src/schema_config_provider/cluster_based.rs (hunk 4 above)
Comment (ShiKaiWi) [reply server/src/schema_config_provider/cluster_based.rs:13]: Yes. Now the schema information is not supported to be stored by CeresMeta. The schema configurations will be fetched from `Cluster` in the future.

File: server/src/route/cluster_based.rs (hunk 2 above)
Comment (ShiKaiWi) [reply server/src/route/cluster_based.rs:32]: I have tried, but I find that the Endpoint is defined in the ceresdbproto crate and I can't impl `FromStr` for it.

Comment (ShiKaiWi) [reply server/src/route/cluster_based.rs:32]: I have tried, but I find that the Endpoint is defined in the ceresdbproto crate and I can't impl `FromStr` for it.

Comment (ShiKaiWi) [reply server/src/route/cluster_based.rs:32]: I have tried, but I find that the Endpoint is defined in the ceresdbproto crate and I can't impl `FromStr` for it.

Comment (ShiKaiWi) [reply server/src/route/cluster_based.rs:32]: I have tried, but I find that the Endpoint is defined in the ceresdbproto crate and I can't impl `FromStr` for it.

Comment (waynexia) [reply server/src/route/cluster_based.rs:32]: I have a suggestion: redefine `Endpoint` in our lib and operate on it, and implement methods like `to_pb()` and `from_pb()`. By doing this https://github.com/CeresDB/ceresdb/pull/215#discussion_r954466140 can also be solved.

**Review by waynexia — APPROVED** [review]

**Pull request merged**